### PR TITLE
ENH: Adding action for execution of virsorter2 run command

### LINKED ISSUE #5: ENH: Add virsorter2-analysis action

### PR DESCRIPTION
**Pull request by ChristosMatzoros**
This PR closes #5 .

- Adds _virsorter2_analysis action which executes virsorter2 main function for viral identification
- Adds a pipeline virsorter2_run which uses _virsorter2_analysis. In a following PR another action will use the tables created by _virsorter2_analysis to create a visualization.

### Set up the environment
```shell
mamba create -n q2-viromics2 -c conda-forge -c bioconda -c https://packages.qiime2.org/qiime2/2023.5/core/passed/  -c defaults q2-types numpy=1.23.5 q2cli virsorter=2 "python=3.8" scikit-learn=0.22.1 pandas seaborn hmmer==3.3 prodigal=2.6 screed=1 ruamel.yaml click pip last ncbi-genome-download checkv pyhmmer
```

###  Activate q2-viromics environment
```shell
conda activate q2-viromics
```

### Clone the repo and checkout the PR branch
```bash
git clone https://github.com/bokulich-lab/q2-viromics.git
cd q2-viromics
git fetch origin pull/6/head:pr-6
git checkout pr-6
pip install -e .
```

#### Download the input [datasets](https://polybox.ethz.ch/index.php/s/zjiyGywgMjLnD7y)
```bash
qiime viromics virsorter2-analysis --i-sequences input.qza --i-database db.qza  --output-dir outDir --verbose
```

### CODE REVIEW

**Comment by VinzentRisch** [comment]
Use this channel 
https://packages.qiime2.org/qiime2/2023.5/staged/core/

File: q2_viromics/virsorter2_run.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        vs2_run_execution(tmp, sequences, database, n_jobs)
+
+        viral_combined_path = os.path.join(tmp, "final-viral-combined.fa")
+        shutil.copy(viral_combined_path, os.path.join(str(viral_sequences)))
+
+        viral_score_fp = os.path.join(tmp, "final-viral-score.tsv")
+        viral_score_df = pd.read_csv(viral_score_fp, sep="\t")
+
+        viral_boundary_fp = os.path.join(tmp, "final-viral-boundary.tsv")
+        viral_boundary_df = pd.read_csv(viral_boundary_fp, sep="\t")
```
Comment (VinzentRisch) [inline q2_viromics/virsorter2_run.py]: Can you please add comments to these steps?

File: q2_viromics/plugin_setup.py [hunk 2]
```diff
@@ -27,15 +37,33 @@
...
     Virsorter2Db,
     directory_format=Virsorter2DbDirFmt,
-    description=("Represents a group Virsorter2 database."),
+    description=("Represents a VirSorter2 database."),
+)
+
+plugin.register_artifact_class(
+    FeatureData[ViralScore],
+    directory_format=ViralScoreDirFmt,
+    description=("Represents viral score table from VirSorter2."),
```
Comment (VinzentRisch) [inline q2_viromics/plugin_setup.py]:
```suggestion
    description=("viral score table from VirSorter2."),
```

File: q2_viromics/plugin_setup.py [hunk 3]
```diff
@@ -27,15 +37,33 @@
...
 )
 
+plugin.register_artifact_class(
+    FeatureData[ViralBoundary],
+    directory_format=ViralBoundaryDirFmt,
+    description=("Represents viral boundary table from VirSorter2."),
```
Comment (VinzentRisch) [inline q2_viromics/plugin_setup.py]:
```suggestion
    description=("Viral boundary table from VirSorter2."),
```

File: q2_viromics/plugin_setup.py [hunk 4]
```diff
@@ -27,15 +37,33 @@
...
 )
-plugin.register_semantic_types(Virsorter2Db)
+
+plugin.register_semantic_types(Virsorter2Db, ViralScore)
 
 plugin.register_artifact_class(
     Virsorter2Db,
     directory_format=Virsorter2DbDirFmt,
-    description=("Represents a group Virsorter2 database."),
+    description=("Represents a VirSorter2 database."),
```
Comment (VinzentRisch) [inline q2_viromics/plugin_setup.py]:
```suggestion
    description=("VirSorter2 database."),
```

File: q2_viromics/types/_transformer.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+from q2_viromics.types._format import ViralBoundaryDirFmt, ViralScoreDirFmt
+
+from ..plugin_setup import plugin
+
+
+@plugin.register_transformer
+def _1(data: pd.DataFrame) -> ViralScoreDirFmt:
+    ff = ViralScoreDirFmt()
+    ViralScore_file_path = os.path.join(str(ff), "final-viral-score.tsv")
+    data.to_csv(ViralScore_file_path, sep="\t", index=False)
```
Comment (VinzentRisch) [inline q2_viromics/types/_transformer.py]:
```suggestion
    data.to_csv(os.path.join(str(ff), "final-viral-score.tsv", sep="\t", index=False)
```

File: q2_viromics/types/_transformer.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+def _2(data: ViralScoreDirFmt) -> pd.DataFrame:
+    file = pd.read_csv(os.path.join(str(data), "final-viral-score.tsv"), sep="\t")
+    return file
+
+
+@plugin.register_transformer
+def _3(data: pd.DataFrame) -> ViralBoundaryDirFmt:
+    ff = ViralBoundaryDirFmt()
+    ViralScore_file_path = os.path.join(str(ff), "final-viral-boundary.tsv")
+    data.to_csv(ViralScore_file_path, sep="\t", index=False)
```
Comment (VinzentRisch) [inline q2_viromics/types/_transformer.py]:
```suggestion
    data.to_csv(os.path.join(str(ff), "final-viral-boundary.tsv", sep="\t", index=False)
```

File: q2_viromics/tests/test_transformers.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+from q2_viromics.types._format import ViralBoundaryDirFmt, ViralScoreDirFmt
+
+
+class TestTransformers(TestPluginBase):
+    package = "q2_viromics.tests"
+
+    def setUp(self):
+        super().setUp()
+
```
Comment (VinzentRisch) [inline q2_viromics/tests/test_transformers.py:24]: define the df here so you dont have to define it in every test.
 df = pd.DataFrame({"col1": [1, 2], "col2": [3, 4]})

**Review by VinzentRisch — CHANGES_REQUESTED** [review]
Looks good!
I ran the command and it works as expected.
See some suggestions below.

**Comment by ChristosMatzoros** [comment]
Hey @VinzentRisch,

I did the changes, you can have a look and let me know.

Thanks

**Review by VinzentRisch — APPROVED** [review]
Looks good!

**Pull request merged**